### ISSUE TITLE
MobileAdvert object issues...

### ISSUE THREAD

**Issue by Bootfit** [Scirra/Construct-3-bugs#1144]
## Problem description

* When adding MobileAdvert object to app, the app then hangs when previewed via remote preview on iPhone 6S+. Also hangs on browser preview (although the attached .c3p file doesn't). Also fails to build via PhoneGapBuild service when exported as iOS(Cordova) when MobileAdvert object is present*

## Attach a .c3p
[MobileAdvertObject.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/1607704/MobileAdvertObject.c3p.zip)

*A minimal Construct 3 project (.c3p) is required to be attached. Please see the guidelines*

## Steps to reproduce

1. Open file
2. Preview as mobile preview
3. Try to build via PhoneGap Build

## Observed result

*What do you see happen?*

See attached .png of iPhone remote preview screen - phone just hangs on Loading...

![img_3094](https://user-images.githubusercontent.com/27493296/34625124-7e05a2fe-f24f-11e7-82d4-dbcd9d7ab3a2.PNG)

## Expected result

*Expected the app to run or to be built*

## Affected browsers

* Chrome: yes
* Firefox: yes/no - ?
* Edge: yes/no - ?
* Safari: yes in iOS

## System details

*If you see a crash report dialog, please copy and paste it to where it says "PASTE HERE" below. Otherwise please go to **Menu** -> **About** -> **Platform information** and paste that information there instead.*

<details><summary>View details</summary>

Platform information
Browser: Chrome
Browser version: 63.0.3239.84
Browser engine: Blink
Browser architecture: (unknown)
Context: browser
Operating system: Mac OS X
Operating system version: 10.13_2
Operating system architecture: (unknown)
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 4
Approx. device memory: 8 GB
User agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.84 Safari/537.36
C3 release: r75
Language setting: en-US

Local storage
Storage quota (approx): 30 gb
Storage usage (approx): 156 mb (0.5%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Vendor: Intel Inc.
Renderer: Intel HD Graphics 4000 OpenGL Engine
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 255.875
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 44100 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
</details>

**Comment by shortercode** [comment]
I believe a server issue is affecting remote preview; specifically for iOS. Due to a safari bug, if the host is on the same network as the client a remote preview connection cannot be made without an additional server ( which isn't currently responding). 

Until that's working again let's look at the other 2 issues. How did you replicate the situation with local preview?

I don't have a PhoneGapBuild account to test with, and I don't believe we guarantee support for it. I believe you mentioned on the forum thread that you had been in contact with their support? Our mobile advert plugin is not hosted on GitHub and I'm not aware of method of adding NPM documentation without linking to a repository. However, I can tell you that our iOS version is written in **Swift** using the **cordova-plugin-add-swift-support** plugin and we include the **Google-Mobile-Ads-SDK**  from cocoapods using cordovas usual dependency resolution method. I expect the issue is related to one or more of those things.

**Comment by shortercode** [comment]
Okay the server issue is resolved, the project now correctly loads over remote preview.

**Comment by Bootfit** [comment]
OK - please bear with me answering some of these - I'm new to all this and learning as I go...

I've just checked on my Mac on Google Chrome - the remote preview of the MobileAdvertObject.c3p I attached still isn't working on my iPhone6S+ Safari.

In regards to NPM documentation, there's documentation for another Scirra plugin here: https://www.npmjs.com/package/construct-mobile-purchase - don't know if this helps anywhere.

In regards to the 'cordova-plugin-add-swift-support' and 'Google-Mobile-Ads-SDK' - I'm not sure what you're suggesting here - is this something I can use instead of the MobileAdvert plugin? However this is resolved you're going to have to walk me through this like an 8 year old as I'm not as clued up on this as you are.

Let me know what you need me to do next to help resolve this.

**Comment by shortercode** [comment]
That's fine, everyone starts somewhere. I'll try to tone back a little :)

So I tested the project you gave me via remote preview on an iPad running on iOS 11 and the project loaded correctly. Does remote preview work for you with different projects on that device? Bear in mind that you need iOS 11 to use remote preview, so it could be that your device is on an older version?

The documentation for a module on NPM is automatically taken from the GitHub page for that module. However, for the mobile-advert-plugin we don't use GitHub so it doesn't have any documentation. Given that the module is only intended to be used with C3 I didn't see the lack of documentation to be a problem at the time.

It wasn't to suggest an alternative, I thought that if you were in contact with the PhoneGapBuild team you could pass that information along to see if any of those details prevented it from being build-able on their service.

As a reminder this plugin is only intended to work as iOS and Android applications, so running it on Chrome or Safari it will not do anything ( but the rest of the project is expected to work correctly ).

**Comment by Bootfit** [comment]
OK - I'll get onto the PhoneGapBuild team.

However I've just tried to remote view the application again in my phone browser and now get a javascript error - see attached image.

![img_3106](https://user-images.githubusercontent.com/27493296/34725044-bb2a1fde-f547-11e7-8f80-fdcab44474f0.jpg)

I'll hold fire contacting the PhoneGap Build team as this error may stop the app being built in PG.

Let me know if you need me to do anything else.

**Comment by shortercode** [comment]
That error is caused by using the CreateBanner action in a non-app environment, it shouldn't affect PG. I've added some more checks to the plugin so that it will either silently fail or trigger one of the error conditions. These changes will go out in the next release.

**Comment by Bootfit** [comment]
OK - I've just posted in the Adobe PhoneGap Build forum and I'll wait to see if there's a response from them - they can be hit and miss when replying.

[https://forums.adobe.com/message/10090989#10090989](url)

I'll also paste here the error that's brought up when building via PhoneGap to see if it means anything to you:

`Check dependencies
“Use Legacy Swift Language Version” (SWIFT_VERSION) is required to be configured correctly for targets which use Swift. Use the [Edit > Convert > To Current Swift Syntax…] menu to choose a Swift version or use the Build Settings editor to configure the build setting directly.
“Use Legacy Swift Language Version” (SWIFT_VERSION) is required to be configured correctly for targets which use Swift. Use the [Edit > Convert > To Current Swift Syntax…] menu to choose a Swift version or use the Build Settings editor to configure the build setting directly.

** ARCHIVE FAILED **


The following build commands failed:
	Check dependencies
(1 failure)
Error code 65 for command: xcodebuild with args: -xcconfig,/private/project/cordova/build-release.xcconfig,-workspace,LiveForTheDing.xcworkspace,-scheme,LiveForTheDing,-configuration,Release,-destination,generic/platform=iOS,-archivePath,LiveForTheDing.xcarchive,archive,CONFIGURATION_BUILD_DIR=/private/project/build/device,SHARED_PRECOMPS_DIR=/private/project/build/sharedpch`

**Comment by Bootfit** [comment]
I see this has been flagged as an external issue now.

Is it not a Construct issue now?

**Comment by shortercode** [comment]
Well we're waiting on PhoneGapBuild to say why it doesn't work on their platform, hence external. The remote preview issue should be fixed in r77.

**Comment by shortercode** [comment]
@Bootfit I'm guessing there's no news related to this? I'm guessing your reason for using Phonegap is not having a Mac?

**Comment by Bootfit** [comment]
@shortercode - No I've heard no news from Adobe - their support is very hit and miss.

I do have a Mac - it's whatI'm currently working on - why would that be/not be an issue?

**Comment by shortercode** [comment]
I was wondering why you were using PhoneGapBuild over our own build service, if there's any improvements you'd like to see I'm the guy to ask.

**Comment by Bootfit** [comment]
I'm using PhoneGap as my app has a native social share plugin (https://github.com/EddyVerbruggen/SocialSharing-PhoneGap-Plugin) that requires the app to be built via PhoneGap. But if I also want to include the Construct MobileAdvert object the PhoneGap build blows out.

I don't think Construct has native social share functionality (just browser-based social login), which is why I opted to go down the PhoneGap route.

To get the mobile adverts working I've now bought a plugin from the Scirra store.

If there are better ways to build etc other than PhoneGap please let me know.

Cheers.

**Comment by shortercode** [comment]
It says on it's README page that it's compatible with cordova as well. I admit that it's not on the build server whitelist at the moment. But I'm happy to review cordova plugins to be added to the whitelist. Are you including it via C3 plugin or are you just adding it to the config.xml?

It's also possible to build an app directly from the cordova project on your Mac, although it's a bit more fiddly.

**Comment by Bootfit** [comment]
In regards to the Social Share plugin - I'm exporting as an iOS Cordova project then editing the config.xml, index.html and adding files in a 'platform' folder as per the instructions. All works fine and builds in PhoneGap, until I then try the exact same above but with the MobileAdvert object included.

To be honest - this is my first app and I'm still feeling my way around in regards to how they're built. I've no idea how I'm even meant to get the .ipa file to tie it in with my developer account or get it to Apple yet. I'm hoping the C3 manual covers some of this also.

**Comment by shortercode** [comment]
I expect something in the mobile-advert plugin just doesn't agree with PhoneGapBuild, it would be good to guarantee compatibility with it but at the moment I don't know what limitations they put on plugins so it's hard to say why.

Most of the application configuration is done through iTunes connect, and you can upload your IPA file to there from **xcode** or a program called **application loader**. This can only be done on a mac, because apple are very protective of their ecosystem.

**Comment by Bootfit** [comment]
Cheers for that - is there more detailed app configuration documentation in the C3 manual?

**Comment by shortercode** [comment]
In general we don't write documentation for third party software, as they have their own documentation. You should be able to find plenty of information online about making iOS applications. The tutorial that I'm writing will cover the process, it will be published to construct.net when it's done.

**Comment by shortercode** [comment]
I haven't updated this bug in awhile. I believe @Bootfit has stopped using the MobileAds plugin so we don't have anyone to track this issue with. If somebody else is having issues with PhoneGapBuild and the MobileAds plugin please open a new issue and I will take a look.

Thanks for your time BTW @Bootfit 

Incidentally the tutorial mentioned in this issue can be found [here](https://www.construct.net/gb/tutorials/ios-application-development-and-publishing-with-xcode-25).
